File: leetcode/0735-asteroid-collision/solution.cpp
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> asteroidCollision(vector<int>& asteroids) {
        vector<int> stack;  // used as a stack via back()/push_back()/pop_back()

        for (int a : asteroids) {
            bool alive = true;

            while (alive && a < 0 && !stack.empty() && stack.back() > 0) {
                if (stack.back() < -a) {
                    stack.pop_back();
                } else if (stack.back() == -a) {
                    stack.pop_back();
                    alive = false;
                } else {
                    alive = false;
                }
            }

            if (alive) {
                stack.push_back(a);
            }
        }

        return stack;
    }
};

File: leetcode/0735-asteroid-collision/solution.py
from typing import List


class Solution:
    def asteroidCollision(self, asteroids: List[int]) -> List[int]:
        stack = []

        for a in asteroids:
            alive = True

            # only a left-mover colliding with a right-mover on top of the stack matters
            while alive and a < 0 and stack and stack[-1] > 0:
                if stack[-1] < -a:
                    stack.pop()  # top explodes, keep checking against what's exposed now
                elif stack[-1] == -a:
                    stack.pop()  # both explode
                    alive = False
                else:
                    alive = False  # incoming asteroid explodes

            if alive:
                stack.append(a)

        return stack
